File: agent/transport/transfer.go
package transport

import (
	"github.com/chriskaliX/Hades/agent/agent"
	"github.com/chriskaliX/Hades/agent/proto"
	"github.com/chriskaliX/Hades/agent/transport/pool"
	"errors"
	"fmt"
	"sync"
	"sync/atomic"
	"time"

	"github.com/chriskaliX/SDK/config"
	"github.com/chriskaliX/SDK/transport/protocol"
	"go.uber.org/zap"
)

var (
	ErrBufferOverflow = errors.New("buffer overflow")
	ErrAgentDataType  = errors.New("agent datatype is not support")
)

// The channel to dispatch to plugin
var (
	PluginTaskChan   = make(chan *proto.Task)
	PluginConfigChan = make(chan map[string]*proto.Config)
)

const size = 8186 // remain 6 space for importance, always available

var Trans = NewTransfer()

type Transfer struct {
	mu         sync.Mutex
	buf        [8192]*proto.Record
	offset     int
	txCnt      uint64
	rxCnt      uint64
	updateTime time.Time
}

func NewTransfer() *Transfer {
	return &Transfer{
		buf:        [8192]*proto.Record{},
		updateTime: time.Now(),
	}
}

// Save the record to the buffer, control the buffer
func (t *Transfer) Transmission(rec *proto.Record, important bool) (err error) {
	t.mu.Lock()
	defer t.mu.Unlock()
	if t.offset >= size {
		if important && t.offset < len(t.buf) {
			t.buf[t.offset] = rec
			t.offset++
		}
		err = ErrBufferOverflow
		return
	}
	t.buf[t.offset] = rec
	t.offset++
	return
}

// TransmissionSDK is an wrapper to satisfy the SDK interface
func (t *Transfer) TransmissionSDK(rec protocol.ProtoType, important bool) (err error) {
	return t.Transmission(rec.(*proto.Record), important)
}

// Send the record from buffer
func (t *Transfer) Send(client proto.Transfer_TransferClient) (err error) {
	// use lock carefully, unlock the field if we need
	t.mu.Lock()
	if t.offset == 0 {
		t.mu.Unlock()
		return
	}
	recs := make([]*proto.Record, t.offset)
	copy(recs, t.buf[:t.offset]) // copy, for reference
	t.offset = 0
	t.mu.Unlock()
	if err = client.Send(&proto.PackagedData{
		Records:      recs,
		AgentId:      agent.ID,
		IntranetIpv4: agent.PrivateIPv4.Load().([]string),
		IntranetIpv6: agent.PrivateIPv6.Load().([]string),
		ExtranetIpv4: agent.PublicIPv4.Load().([]string),
		ExtranetIpv6: agent.PublicIPv6.Load().([]string),
		Hostname:     agent.Hostname.Load().(string),
		Version:      agent.Version,
		Product:      agent.Product,
	}); err == nil {
		atomic.AddUint64(&t.txCnt, uint64(len(recs)))
	}
	for _, rec := range recs {
		pool.Put(rec)
	}
	return
}

func (t *Transfer) Receive(client proto.Transfer_TransferClient) (err error) {
	cmd, err := client.Recv()
	if err != nil {
		return
	}
	// precheck cmd
	if cmd == nil {
		return
	}
	zap.S().Info("command received")
	atomic.AddUint64(&t.rxCnt, 1)
	// resolve task & config
	t.resolveTask(cmd)
	agent.SetRunning()
	t.resolveConfig(cmd)
	return
}

func (t *Transfer) GetState(now time.Time) (txTPS, rxTPS float64) {
	instant := now.Sub(t.updateTime).Seconds()
	if instant != 0 {
		txTPS = float64(atomic.SwapUint64(&t.txCnt, 0)) / float64(instant)
		rxTPS = float64(atomic.SwapUint64(&t.rxCnt, 0)) / float64(instant)
	}
	t.updateTime = now
	return
}

func (t *Transfer) resolveConfig(cmd *proto.Command) (err error) {
	if cmd.Configs == nil {
		return
	}
	configs := map[string]*proto.Config{}
	for _, config := range cmd.Configs {
		configs[config.Name] = config
	}
	if config, ok := configs[agent.Product]; ok && config.Version != agent.Version {
		zap.S().Infof("agent will update from version %v to version %v", agent.Version, config.Version)
		if err = agent.Update(*config); err == nil {
			zap.S().Info("agent update successfully")
			agent.Cancel()
			return
		}
		zap.S().Errorf("agent update failed: %s", err.Error())
		agent.SetAbnormal(fmt.Sprintf("agent update failed: %s", err.Error()))
	}
	delete(configs, agent.Product)
	select {
	case PluginConfigChan <- configs:
	default:
		err = fmt.Errorf("plugin configchan is syncing or is cancelled")
		zap.S().Error(err.Error())
	}
	return
}

func (t *Transfer) resolveTask(cmd *proto.Command) (err error) {
	if cmd.Task == nil {
		return
	}
	if cmd.Task.ObjectName == agent.Product {
		switch cmd.Task.DataType {
		// according to the service, restart will be 45 secs in systemd based daemon
		// crontab is used in sysvinit to keep the agent always available
		case config.TaskShutdown, config.TaskRestart:
			zap.S().Info("agent shutdown is called")
			TaskSuccess(cmd.Task.Token, "agent shutdown is called")
			agent.Cancel()
			return
		case config.TaskSetenv:
		default:
			TaskError(cmd.Task.Token, fmt.Sprintf("agent datatype %d is not supported", cmd.Task.DataType))
			return ErrAgentDataType
		}
	}
	select {
	case PluginTaskChan <- cmd.Task:
	default:
		err = fmt.Errorf("plugin taskchan is syncing or is cancelled")
		TaskError(cmd.Task.Token, err.Error())
	}
	return
}

func TaskSuccess(token string, msg string) {
	zap.S().Info(token, msg)
	Trans.Transmission(&proto.Record{
		DataType:  5100,
		Timestamp: time.Now().Unix(),
		Data: &proto.Payload{
			Fields: map[string]string{
				"token":  token,
				"msg":    msg,
				"status": "success",
			},
		},
	}, true)
}

func TaskError(token string, msg string) {
	zap.S().Error(token, msg)
	Trans.Transmission(&proto.Record{
		DataType:  5100,
		Timestamp: time.Now().Unix(),
		Data: &proto.Payload{
			Fields: map[string]string{
				"token":  token,
				"msg":    msg,
				"status": "fail",
			},
		},
	}, true)
}

File: agent/src/transport/transfer.rs
use std::{
    collections::HashMap,
    sync::{
        atomic::{AtomicU64, AtomicUsize, Ordering},
        Arc, LazyLock, OnceLock,
    },
};

use coarsetime::{Clock, Instant};

use parking_lot::Mutex;
use prost::Message as _;
use tokio::sync::mpsc;
use tokio::sync::mpsc::error::TrySendError;

use crate::{
    agent,
    proto::{Config, PackagedData, Payload, Record, Task, TaskType},
};
use tokio_util::sync::CancellationToken;

pub const BUFFER_TOTAL:  usize = 8192;
pub const BUFFER_NORMAL: usize = 8160;

pub static PLUGIN_TASK_CHAN:   OnceLock<mpsc::Sender<Task>>                    = OnceLock::new();
pub static PLUGIN_CONFIG_CHAN: OnceLock<mpsc::Sender<HashMap<String, Config>>> = OnceLock::new();

static TRANS: LazyLock<Transfer> = LazyLock::new(Transfer::new);

/// Returns a reference to the process-global [`Transfer`] instance.
pub fn trans() -> &'static Transfer { &TRANS }

/// Pre-built, non-records portion of a [`PackagedData`] frame.
/// Rebuilt only when `agent::host()` issues a new [`Arc`] (IP/hostname change).
struct HdrCache {
    snapshot:      Arc<crate::agent::host::HostInfo>,
    agent_id:      String,
    version:       String,
    product:       String,
    hostname:      String,
    intranet_ipv4: Vec<String>,
    intranet_ipv6: Vec<String>,
    extranet_ipv4: Vec<String>,
    extranet_ipv6: Vec<String>,
}

impl HdrCache {
    fn build(h: Arc<crate::agent::host::HostInfo>) -> Self {
        Self {
            agent_id:      agent::ID.clone(),
            version:       agent::VERSION.to_owned(),
            product:       agent::PRODUCT.to_owned(),
            hostname:      h.hostname.clone(),
            intranet_ipv4: nonempty_vec(&h.private_ipv4),
            intranet_ipv6: nonempty_vec(&h.private_ipv6),
            extranet_ipv4: nonempty_vec(&h.public_ipv4),
            extranet_ipv6: nonempty_vec(&h.public_ipv6),
            snapshot:      h,
        }
    }
}

pub struct Transfer {
    pub(super) inner:       Mutex<Vec<Record>>,
    pub        tx_cnt:      AtomicU64,
    pub        rx_cnt:      AtomicU64,
    /// Records dropped due to buffer overflow.
    pub        drop_cnt:    AtomicU64,
    /// Approximate buffer occupancy; allows a fast idle-skip without locking.
    pub        len_hint:    AtomicUsize,
               update_time: Mutex<Instant>,
    /// Cached PackagedData header; rebuilt only on host-info change.
               hdr:         Mutex<HdrCache>,
}

impl Transfer {
    pub(super) fn new() -> Self {
        Self {
            inner:       Mutex::new(Vec::with_capacity(BUFFER_TOTAL)),
            tx_cnt:      AtomicU64::new(0),
            rx_cnt:      AtomicU64::new(0),
            drop_cnt:    AtomicU64::new(0),
            update_time: Mutex::new(Instant::now()),
            len_hint:    AtomicUsize::new(0),
            hdr:         Mutex::new(HdrCache::build(agent::host::get())),
        }
    }

    pub fn transmission(&self, rec: Record, important: bool) -> anyhow::Result<()> {
        let mut buf = self.inner.lock();
        if buf.len() >= BUFFER_NORMAL {
            if important && buf.len() < BUFFER_TOTAL {
                buf.push(rec);
                self.len_hint.fetch_add(1, Ordering::Relaxed);
                return Ok(());
            }
            self.drop_cnt.fetch_add(1, Ordering::Relaxed);
            anyhow::bail!("buffer overflow");
        }
        buf.push(rec);
        self.len_hint.fetch_add(1, Ordering::Relaxed);
        Ok(())
    }

    /// Swap the internal buffer with `out` (which must be empty and pre-allocated).
    /// After the call `out` holds all buffered records; the internal buffer
    /// reclaims `out`'s allocation so producers never block on a re-alloc.
    pub fn drain_into(&self, out: &mut Vec<Record>) {
        debug_assert!(out.is_empty());
        let mut buf = self.inner.lock();
        std::mem::swap(&mut *buf, out);
        self.len_hint.store(0, Ordering::Relaxed);
    }

    /// Atomically drain the record buffer and send one [`PackagedData`] frame.
    ///
    /// Mirrors Go's `Transfer.Send(client)`: lock → drain → unlock → build → send.
    /// Returns `false` if the channel is closed (session should terminate),
    /// `true` in all other cases (including when the buffer was empty).
    pub async fn send(&self, tx: &mpsc::Sender<PackagedData>) -> bool {
        // Fast path: skip mutex acquisition entirely when nothing is buffered.
        if self.len_hint.load(Ordering::Relaxed) == 0 { return true; }
        let mut records = Vec::new();
        self.drain_into(&mut records);
        if records.is_empty() { return true; }
        let count = records.len() as u64;
        // Use cached header; rebuild only when host info changes (rare).
        let msg = {
            let mut hdr = self.hdr.lock();
            let h = agent::host::get();
            if !Arc::ptr_eq(&h, &hdr.snapshot) {
                *hdr = HdrCache::build(h);
            }
            PackagedData {
                records,
                payloads:      vec![],
                agent_id:      hdr.agent_id.clone(),
                intranet_ipv4: hdr.intranet_ipv4.clone(),
                intranet_ipv6: hdr.intranet_ipv6.clone(),
                extranet_ipv4: hdr.extranet_ipv4.clone(),
                extranet_ipv6: hdr.extranet_ipv6.clone(),
                hostname:      hdr.hostname.clone(),
                version:       hdr.version.clone(),
                product:       hdr.product.clone(),
            }
        };
        let byte_len = msg.encoded_len() as u64;
        if tx.send(msg).await.is_err() { return false; }
        self.tx_cnt.fetch_add(count, Ordering::Relaxed);
        crate::transport::connection::stats_handler()
            .tx_bytes.fetch_add(byte_len, Ordering::Relaxed);
        true
    }

    pub fn get_state(&self) -> (f64, f64) {
        let now = Instant::now();
        let mut ut = self.update_time.lock();
        let secs = (now - *ut).as_f64();
        *ut = now;
        if secs > 0.0 {
            let tx = self.tx_cnt.swap(0, Ordering::Relaxed) as f64 / secs;
            let rx = self.rx_cnt.swap(0, Ordering::Relaxed) as f64 / secs;
            (tx, rx)
        } else {
            (0.0, 0.0)
        }
    }
}

fn nonempty_vec(s: &str) -> Vec<String> {
    (!s.is_empty()).then(|| s.to_owned()).into_iter().collect()
}

fn task_record(token: &str, msg: &str, status: &str) -> Record {
    let ts = Clock::now_since_epoch().as_secs() as i64;
    Record {
        data_type: 5100,
        timestamp: ts,
        data: Some(Payload {
            fields: HashMap::from([
                ("token".into(),  token.into()),
                ("msg".into(),    msg.into()),
                ("status".into(), status.into()),
            ]),
        }),
    }
}

pub fn task_success(token: &str, msg: &str) {
    log::info!("task success: {token} {msg}");
    let _ = trans().transmission(task_record(token, msg, "success"), true);
}

pub fn task_error(token: &str, msg: &str) {
    log::error!("task error: {token} {msg}");
    let _ = trans().transmission(task_record(token, msg, "fail"), true);
}

pub fn resolve_task(task: Task, ct: &CancellationToken) -> anyhow::Result<()> {
    if task.object_name == agent::PRODUCT {
        match TaskType::try_from(task.data_type) {
            Ok(TaskType::TaskShutdown) | Ok(TaskType::TaskRestart) => {
                log::info!("agent shutdown/restart requested");
                task_success(&task.token, "agent shutdown is called");
                ct.cancel();
                return Ok(());
            }
            Ok(TaskType::TaskSetenv | TaskType::TaskUnsetenv) => {
                // env ops are forwarded to plugins; fall through.
            }
            Ok(TaskType::TaskUpload) => {
                #[derive(serde::Deserialize)]
                struct UploadPayload { path: String, #[serde(default)] buf_size: u64 }
                match serde_json::from_str::<UploadPayload>(&task.data) {
                    Ok(p) => {
                        let req = super::file::UploadRequest {
                            path:     p.path,
                            buf_size: p.buf_size,
                            token:    task.token.clone(),
                        };
                        if let Err(e) = super::file::upload_file(req) {
                            task_error(&task.token, &e.to_string());
                        }
                    }
                    Err(e) => task_error(&task.token, &format!("upload: bad payload: {e}")),
                }
                return Ok(());
            }
            _ => {
                let msg = format!("agent datatype {} is not supported", task.data_type);
                task_error(&task.token, &msg);
                anyhow::bail!(msg);
            }
        }
    }
    if let Some(sender) = PLUGIN_TASK_CHAN.get() {
        match sender.try_send(task.clone()) {
            Ok(()) => {}
            Err(TrySendError::Full(_)) | Err(TrySendError::Closed(_)) => {
                let msg = "plugin task channel is syncing or is cancelled";
                task_error(&task.token, msg);
                anyhow::bail!(msg);
            }
        }
    } else {
        let msg = "plugin task channel is not initialized";
        task_error(&task.token, msg);
        anyhow::bail!(msg);
    }
    Ok(())
}

pub async fn resolve_config(configs: Vec<Config>, ct: &CancellationToken) -> anyhow::Result<()> {
    let mut map: HashMap<String, Config> = configs
        .into_iter()
        .map(|c| (c.name.clone(), c))
        .collect();

    if let Some(cfg) = map.get(agent::PRODUCT) {
        if cfg.version != agent::VERSION {
            log::info!("agent update {} → {}", agent::VERSION, cfg.version);
            let update_cfg = agent::update::UpdateConfig {
                sha256:        cfg.sha256.clone(),
                download_urls: cfg.download_urls.clone(),
                pkg_type:      cfg.r#type.clone(),
            };
            let token_clone = ct.clone();
            let result = tokio::task::spawn_blocking(move || agent::update::update(&token_clone, &update_cfg))
                .await
                .map_err(|e| anyhow::anyhow!("spawn_blocking join: {e}"))?;

            match result {
                Ok(_) => {
                    log::info!("agent update successful");
                    ct.cancel();
                    return Ok(());
                }
                Err(e) => {
                    let msg = format!("agent update failed: {e}");
                    log::error!("{msg}");
                    agent::state::set_abnormal(&msg);
                }
            }
        }
    }

    map.remove(agent::PRODUCT);
    if let Some(sender) = PLUGIN_CONFIG_CHAN.get() {
        match sender.try_send(map) {
            Ok(()) => {}
            Err(TrySendError::Full(_)) | Err(TrySendError::Closed(_)) => {
                anyhow::bail!("plugin config channel is syncing or is cancelled");
            }
        }
    } else if !map.is_empty() {
        anyhow::bail!("plugin config channel is not initialized");
    }
    Ok(())
}
